api/src/main/java/org/apache/iceberg/RowIdentifier.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+
+/**
+ * Row identifier of a table.
+ * <p>
+ * Row identifier is a definition of table row uniqueness,
+ * similar to the concept of primary key in a relational database system.
+ * A row should be unique in a table based on the values of each {@link RowIdentifyField}.
+ * Iceberg itself does not enforce row uniqueness based on this identifier.
+ * It is leveraged by operations such as streaming upsert.
+ */
+public class RowIdentifier implements Serializable {
```

Comment:
I have considered some alternatives for the name, and discarded the following:
  1. primary key: used in db systems and implies enforcing uniqueness, which does not fit Iceberg use case
  2. upsert id: too specific for the upsert use case
  3. row id: conflicts with `java.sql.RowId`, and gives a feeling of referring to a specific row
  4. default row id: defaultXxx is used a lot in metadata, having a class with a default in prefix makes code in table metadata confusing.

Reply:
`RowKey` sounds good to me, I will update based on that!